### PR TITLE
external: create a configmap for the command line args

### PR DESCRIPTION
**Pull request by parth-gr**
user can look back to there configurations by
looking at the configmap created with command
line arguments
This will be useful for them during upgrades when they need to re run the python script with the same flags

**Checklist:**

- [ ] **Commit Message Formatting**: Commit titles and messages follow guidelines in the [developer guide](https://rook.io/docs/rook/latest/Contributing/development-flow/#commit-structure).
- [ ] Reviewed the developer guide on [Submitting a Pull Request](https://rook.io/docs/rook/latest/Contributing/development-flow/#submitting-a-pull-request)
- [ ] [Pending release notes](https://github.com/rook/rook/blob/master/PendingReleaseNotes.md) updated with breaking and/or notable changes for the next minor release.
- [ ] Documentation has been updated, if necessary.
- [ ] Unit tests have been added, if necessary.
- [ ] Integration tests have been added, if necessary.

### CODE REVIEW

**Comment by parth-gr** [comment]
I have only added the arguments that has some value
```
sh-4.4$ python3 a.py --rbd-data-pool-name replicapool
[{"name": "external-cluster-user-command", "kind": "ConfigMap", "data": {"data": ["verbose: False", "ceph_conf: None", "keyring: None", "rgw_pool_prefix: default", "restricted_auth_permission: False", "v2_port_enable: False", "format: json", "cephfs_filesystem_name: myfs", "cephfs_metadata_pool_name: myfs-metadata", "cephfs_data_pool_name: myfs-replicated", "rbd_data_pool_name: replicapool", "rgw_skip_tls: False", "skip_monitoring_endpoint: False", "dry_run: False", "upgrade: False"]}}, {"name": "rook-ceph-mon-endpoints", "kind": "ConfigMap", "data": {"data": "a=10.97.22.211:6789", "maxMonId": "0", "mapping": "{}"}}, {"name": "rook-ceph-mon", "kind": "Secret", "data": {"admin-secret": "admin-secret", "fsid": "cacf16f1-a616-4fa7-abed-e0b373739a12", "mon-secret": "mon-secret"}}, {"name": "rook-ceph-operator-creds", "kind": "Secret", "data": {"userID": "client.healthchecker", "userKey": "AQAgf3Jmrz0SGBAAdRNMCORxP6tvQA80QX9jUQ=="}}, {"name": "monitoring-endpoint", "kind": "CephCluster", "data": {"MonitoringEndpoint": "10.244.239.85", "MonitoringPort": "9283"}}, {"name": "rook-csi-rbd-node", "kind": "Secret", "data": {"userID": "csi-rbd-node", "userKey": "AQBDcnJmycs5ExAAWK+bf6EQ5mjOhNUHo6cjOQ=="}}, {"name": "rook-csi-rbd-provisioner", "kind": "Secret", "data": {"userID": "csi-rbd-provisioner", "userKey": "AQBDcnJml+NaBBAAynhQU44b416JvlgRla5Bpw=="}}, {"name": "rook-csi-cephfs-provisioner", "kind": "Secret", "data": {"adminID": "csi-cephfs-provisioner", "adminKey": "AQBDcnJmp8xFJRAA07NNBroRhS9zsSwT/Nw7vA=="}}, {"name": "rook-csi-cephfs-node", "kind": "Secret", "data": {"adminID": "csi-cephfs-node", "adminKey": "AQBDcnJmKaAyNRAAX9u7L0DT2htQjKTwygOHhA=="}}, {"name": "rook-ceph-dashboard-link", "kind": "Secret", "data": {"userID": "ceph-dashboard-link", "userKey": "http://10.244.239.85:7000/"}}, {"name": "ceph-rbd", "kind": "StorageClass", "data": {"pool": "replicapool", "csi.storage.k8s.io/provisioner-secret-name": "rook-csi-rbd-provisioner", "csi.storage.k8s.io/controller-expand-secret-name": "rook-csi-rbd-provisioner", "csi.storage.k8s.io/node-stage-secret-name": "rook-csi-rbd-node"}}, {"name": "cephfs", "kind": "StorageClass", "data": {"fsName": "myfs", "pool": "myfs-replicated", "csi.storage.k8s.io/provisioner-secret-name": "rook-csi-cephfs-provisioner", "csi.storage.k8s.io/controller-expand-secret-name": "rook-csi-cephfs-provisioner", "csi.storage.k8s.io/node-stage-secret-name": "rook-csi-cephfs-node"}}]
```

**Comment by parth-gr** [comment]
> I have only added the arguments that has some value
> 
> ```
> sh-4.4$ python3 a.py --rbd-data-pool-name replicapool
> [{"name": "external-cluster-user-command", "kind": "ConfigMap", "data": {"data": ["verbose: False", "ceph_conf: None", "keyring: None", "rgw_pool_prefix: default", "restricted_auth_permission: False", "v2_port_enable: False", "format: json", "cephfs_filesystem_name: myfs", "cephfs_metadata_pool_name: myfs-metadata", "cephfs_data_pool_name: myfs-replicated", "rbd_data_pool_name: replicapool", "rgw_skip_tls: False", "skip_monitoring_endpoint: False", "dry_run: False", "upgrade: False"]}}, {"name": "rook-ceph-mon-endpoints", "kind": "ConfigMap", "data": {"data": "a=10.97.22.211:6789", "maxMonId": "0", "mapping": "{}"}}, {"name": "rook-ceph-mon", "kind": "Secret", "data": {"admin-secret": "admin-secret", "fsid": "cacf16f1-a616-4fa7-abed-e0b373739a12", "mon-secret": "mon-secret"}}, {"name": "rook-ceph-operator-creds", "kind": "Secret", "data": {"userID": "client.healthchecker", "userKey": "AQAgf3Jmrz0SGBAAdRNMCORxP6tvQA80QX9jUQ=="}}, {"name": "monitoring-endpoint", "kind": "CephCluster", "data": {"MonitoringEndpoint": "10.244.239.85", "MonitoringPort": "9283"}}, {"name": "rook-csi-rbd-node", "kind": "Secret", "data": {"userID": "csi-rbd-node", "userKey": "AQBDcnJmycs5ExAAWK+bf6EQ5mjOhNUHo6cjOQ=="}}, {"name": "rook-csi-rbd-provisioner", "kind": "Secret", "data": {"userID": "csi-rbd-provisioner", "userKey": "AQBDcnJml+NaBBAAynhQU44b416JvlgRla5Bpw=="}}, {"name": "rook-csi-cephfs-provisioner", "kind": "Secret", "data": {"adminID": "csi-cephfs-provisioner", "adminKey": "AQBDcnJmp8xFJRAA07NNBroRhS9zsSwT/Nw7vA=="}}, {"name": "rook-csi-cephfs-node", "kind": "Secret", "data": {"adminID": "csi-cephfs-node", "adminKey": "AQBDcnJmKaAyNRAAX9u7L0DT2htQjKTwygOHhA=="}}, {"name": "rook-ceph-dashboard-link", "kind": "Secret", "data": {"userID": "ceph-dashboard-link", "userKey": "http://10.244.239.85:7000/"}}, {"name": "ceph-rbd", "kind": "StorageClass", "data": {"pool": "replicapool", "csi.storage.k8s.io/provisioner-secret-name": "rook-csi-rbd-provisioner", "csi.storage.k8s.io/controller-expand-secret-name": "rook-csi-rbd-provisioner", "csi.storage.k8s.io/node-stage-secret-name": "rook-csi-rbd-node"}}, {"name": "cephfs", "kind": "StorageClass", "data": {"fsName": "myfs", "pool": "myfs-replicated", "csi.storage.k8s.io/provisioner-secret-name": "rook-csi-cephfs-provisioner", "csi.storage.k8s.io/controller-expand-secret-name": "rook-csi-cephfs-provisioner", "csi.storage.k8s.io/node-stage-secret-name": "rook-csi-cephfs-node"}}]
> ```

or do we want flags like `rgw-pool-prefix` instead of `rgw_pool_prefix`?

**Comment by parth-gr** [comment]
@BlaineEXE  For the command output in config map,

Which way we should output:

Lets say command is: python3 a.py --abc 123 --pqr 456
1) The exact command
"cmd": "python3 a.py --abc 123 --pqr 456"

2) Or individual flags,

"abc": 123
"pqr": 456

**Comment by BlaineEXE** [comment]
In order to facilitate user ease of use, I think it makes most sense to store things in a format that is immediately usable without users having to do anything. For storing configs based on CLI flags, I think it makes the most sense to store them in flag form like `--abc 123 --pqr 456`. I think that it probably makes the most sense to record only the flags themselves and not the command.

Assuming work is completed to implement config file support, we can instead output a generated config file, which might be more legible like your second example but still easily usable.

**Comment by parth-gr** [comment]
So I am printing `print_command.py --abc 123 --pqr 456`

**Comment by parth-gr** [comment]
Output: {"name": "external-cluster-user-command", "kind": "ConfigMap", "data": {"data": "--rbd-data-pool-name replicapool"}}

**Comment by BlaineEXE** [comment]
Do we need any upstream doc updates for this?

**Comment by parth-gr** [comment]
> Do we need any upstream doc updates for this?

@BlaineEXE that was a good reminder,

So I even created the configmap in the import script that upstream uses.

And then I realised we haven't document on, if we need to upgrade and utilise new feature we need to re-run the script, so I need to draw a attention first on how to utilise new flag on a brow field cluster, so I am creating a new issue in which I will document the same and also add about the command line args configmap

File: deploy/examples/import-external-cluster.sh [hunk 1]
```diff
@@ -167,6 +168,18 @@ function importConfigMap() {
...
 
+function createInputCommadConfigMap() {
+  if ! $KUBECTL -n "$NAMESPACE" get configmap "$EXTERNAL_COMMAND_CONFIGMAP_NAME" &>/dev/null; then
+    $KUBECTL -n "$NAMESPACE" \
+      create \
+      configmap \
+      "$EXTERNAL_COMMAND_CONFIGMAP_NAME" \
+      --from-literal=command="$EXTERNAL_CLUSTER_USER_COMMAND" \
+      else
+    echo "configmap $EXTERNAL_COMMAND_CONFIGMAP_NAME already exists"
```
Comment (BlaineEXE) [inline deploy/examples/import-external-cluster.sh:179]: If it already exists, should we assume that this is a re-run with an update? In that case I think the tool should update the existing cm

File: deploy/examples/import-external-cluster.sh [hunk 2]
```diff
@@ -167,6 +168,18 @@ function importConfigMap() {
...
 }
 
+function createInputCommadConfigMap() {
+  if ! $KUBECTL -n "$NAMESPACE" get configmap "$EXTERNAL_COMMAND_CONFIGMAP_NAME" &>/dev/null; then
+    $KUBECTL -n "$NAMESPACE" \
+      create \
+      configmap \
+      "$EXTERNAL_COMMAND_CONFIGMAP_NAME" \
+      --from-literal=command="$EXTERNAL_CLUSTER_USER_COMMAND" \
+      else
```
Comment (BlaineEXE) [inline deploy/examples/import-external-cluster.sh:178]:
```suggestion
  else
```

File: deploy/examples/create-external-cluster-resources.py [hunk 3]
```diff
@@ -1528,6 +1528,9 @@ def init_topology_rbd_pools(self, topology_rbd_pools):
         for pool in topology_rbd_pools:
             self.init_rbd_pool(pool)
 
+    def getCommand(self):
```
Comment (BlaineEXE) [inline deploy/examples/create-external-cluster-resources.py:1531]: `getCommand` makes me think this is getting some sort of command itself. Let's use a clearer name like `getScriptCliFlags` so it is clear what the function is doing.

```suggestion
    def getScriptCliFlags(self):
```

File: deploy/examples/import-external-cluster.sh (hunk 1 above)
Comment (parth-gr) [reply deploy/examples/import-external-cluster.sh]: Thats correct we need to patch it.

File: deploy/examples/import-external-cluster.sh [hunk 4]
```diff
@@ -167,6 +168,20 @@ function importConfigMap() {
...
 
+function createInputCommadConfigMap() {
+  if ! $KUBECTL -n "$NAMESPACE" get configmap "$EXTERNAL_COMMAND_CONFIGMAP_NAME" &>/dev/null; then
+    $KUBECTL -n "$NAMESPACE" \
+      create \
+      configmap \
+      "$EXTERNAL_COMMAND_CONFIGMAP_NAME" \
+      --from-literal=command="$EXTERNAL_CLUSTER_USER_COMMAND"
+  else
+    echo "configmap $EXTERNAL_COMMAND_CONFIGMAP_NAME already exists, update it"
```
Comment (BlaineEXE) [inline deploy/examples/import-external-cluster.sh:179]: Users may interpret "update it" to mean that they need to take action. 

```suggestion
    echo "configmap $EXTERNAL_COMMAND_CONFIGMAP_NAME already exists, updating it"
```

File: deploy/examples/import-external-cluster.sh [hunk 5]
```diff
@@ -167,6 +168,20 @@ function importConfigMap() {
...
+    $KUBECTL -n "$NAMESPACE" patch configmap "$EXTERNAL_COMMAND_CONFIGMAP_NAME" \
+      -p "{'data':{'command':$EXTERNAL_COMMAND_CONFIGMAP_NAME}}"
```
Comment (BlaineEXE) [inline deploy/examples/import-external-cluster.sh:181]: Let's try to be consistent with formatting. The `$KUBECTL create X Y --vals` commands elsewhere all use the format below. Let's do the same for this `patch` command to keep consistency.

```
$KUBECTL -n "$NAMESPACE" \
  create \
  X \
  Y \
  --vals
```

**Review by BlaineEXE — CHANGES_REQUESTED** [review]

File: deploy/examples/import-external-cluster.sh (hunk 5 above)
Comment (parth-gr) [reply deploy/examples/import-external-cluster.sh:181]: Done

**Review by BlaineEXE — APPROVED** [review]

**Comment by travisn** [comment]
@parth-gr What upstream documentation can we update so users will be aware of this new feature?

**Comment by travisn** [comment]
> @parth-gr What upstream documentation can we update so users will be aware of this new feature?

Or maybe it's already covered by #14353?

**Pull request merged**